packages/maskbook/src/web3/hooks/useEtherWrapperCallback.ts

```diff
@@ -0,0 +1,161 @@
+                },
+            )
+        },
+        [account, wrapperContract],
+    )
+
+    const unwrapCallback = useCallback(
+        async (all = true, amount = '0') => {
+            if (!wrapperContract || !amount) {
+                setTransactionState({
+                    type: TransactionStateType.UNKNOWN,
+                })
+                return
+            }
+
+            // read balance
+            const wethBalance = await wrapperContract.methods.balanceOf(account).call()
+
+            // error: invalid withdraw amount
+            if (all === false && new BigNumber(amount).isZero()) {
+                setTransactionState({
+                    type: TransactionStateType.FAILED,
+                    error: new Error('Invalid withdraw amount'),
+                })
+                return
+            }
+
+            // error: insufficent weth balance
+            if (all === false && new BigNumber(wethBalance).isLessThan(amount)) {
+                setTransactionState({
+                    type: TransactionStateType.FAILED,
+                    error: new Error('Insufficent WETH balance'),
+                })
+                return
+            }
+
+            // step 1: start waiting for provider to confirm tx
+            setTransactionState({
+                type: TransactionStateType.WAIT_FOR_CONFIRMING,
+            })
+
+            const config: Tx = {
+                from: account,
+            }
+            const withdrawAmount = all ? wethBalance : amount
+
+            // step 2: estimate gas
+            const estimatedGas = await wrapperContract.methods
+                .withdraw(withdrawAmount)
+                .estimateGas(config)
+                .catch((error) => {
+                    setTransactionState({
+                        type: TransactionStateType.FAILED,
+                        error,
+                    })
+                    throw error
+                })
+
+            // step 3: blocking
+            wrapperContract.methods.withdraw(withdrawAmount).send(
+                {
+                    gas: addGasMargin(estimatedGas).toFixed(),
+                    ...config,
+                },
+                (error, hash) => {
+                    if (error) {
+                        setTransactionState({
+                            type: TransactionStateType.FAILED,
+                            error,
+                        })
+                        return
+                    }
+                    setTransactionState({
+                        type: TransactionStateType.HASH,
+                        hash,
+                    })
+                },
+            )
+        },
+        [account, wrapperContract],
+    )
+
+    const resetCallback = useCallback(() => {
+        setTransactionState({
+            type: TransactionStateType.UNKNOWN,
+        })
+    }, [])
+
+    return [transactionState, wrapCallback, unwrapCallback, resetCallback] as const
```

Comment:
Perhaps it would be more beneficial to return an object for deconstruction

Reply:
It will be good for this case. But I made the same rule for the returns of these kinds of hooks. They return stuff like:

```typescript
[
    transactionState,
    actionCallback, // callback for updating the transaction state
    resetCallback // callback for resetting the transaction state
]
```

Sometimes, it has to return more than one `actionCallback`, which looks weird at first glance.